lib/collections.go

```diff
@@ -949,6 +1022,69 @@ func zipLists(arg0, arg1 ref.Val) ref.Val {
 	return types.NewRefValMap(types.DefaultTypeAdapter, m)
 }
 
+func mapKeys(val ref.Val) ref.Val {
+	mapK, ok := val.(traits.Mapper)
+	if !ok {
+		return types.ValOrErr(mapK, "no such overload")
+	}
+	n, ok := mapK.Size().(types.Int)
+	if !ok {
+		return types.NewErr("unable to get size of map")
+	}
+	keys := make([]ref.Val, 0, n)
+	if mapK.Size() != types.IntZero {
+		canSort := true
+		it := mapK.Iterator()
+		for it.HasNext() == types.True {
+			k := it.Next()
+			keys = append(keys, k)
+			_, ok := k.(traits.Comparer)
+			if !ok {
+				canSort = false
+			}
+		}
+		if canSort {
+			sort.Slice(keys, func(i, j int) bool {
+				return keys[i].(traits.Comparer).Compare(keys[j]) == types.Int(-1)
+			})
+		}
+	}
+	return types.NewRefValList(types.DefaultTypeAdapter, keys)
+}
+
+func mapValues(val ref.Val) ref.Val {
+	mapK, ok := val.(traits.Mapper)
+	if !ok {
+		return types.ValOrErr(mapK, "no such overload")
+	}
+	n, ok := mapK.Size().(types.Int)
+	if !ok {
+		return types.NewErr("unable to get size of map")
+	}
+	keys := make([]ref.Val, 0, n)
+	values := make([]ref.Val, 0, n)
+	if mapK.Size() != types.IntZero {
+		canSort := true
+		it := mapK.Iterator()
+		for it.HasNext() == types.True {
+			k := it.Next()
+			v := mapK.Get(k)
+			keys = append(keys, k)
+			values = append(values, v)
+			_, ok := v.(traits.Comparer)
+			if !ok {
+				canSort = false
+			}
+		}
+		if canSort {
+			sort.Slice(values, func(i, j int) bool {
+				return keys[i].(traits.Comparer).Compare(keys[j]) == types.Int(-1)
+			})
```

Comment:
I don't think this is right. You are sorting the values on the basis of keys, but you are not updating the key ordering. You will need to use a `[]struct{k traits.Comparer, v ref.Value}` as your collection and then sort that (or make a `sort.Interface` implementation that considers two slices if you are old-school — I think the former is better).

Reply:
Thanks for making it clear. I updated the test and it did fail with current logic. 
I now updated the sorting logic using struct as suggested.